sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala

```diff
@@ -984,6 +984,97 @@ case class GpuFastSampleExec(
   }
 }
 
+class GpuRangeIterator(
+    partitionStart: BigInt,
+    partitionEnd: BigInt,
+    step: Long,
+    maxRowCountPerBatch: Long,
+    taskContext: TaskContext,
+    opTime: GpuMetric) extends Iterator[ColumnarBatch] with Logging {
+
+  private def getSafeMargin(bi: BigInt): Long = {
+    if (bi.isValidLong) {
+      bi.toLong
+    } else if (bi > 0) {
+      Long.MaxValue
+    } else {
+      Long.MinValue
+    }
+  }
+
+  private val safePartitionStart = getSafeMargin(partitionStart) // inclusive
+  private val safePartitionEnd = getSafeMargin(partitionEnd) // exclusive, unless start == this
+  private[this] var number: Long = safePartitionStart
+  private[this] var done: Boolean = false
+
+  override def hasNext: Boolean = {
+    if (!done) {
+      if (step > 0) {
+        number < safePartitionEnd
+      } else {
+        number > safePartitionEnd
+      }
+    } else false
+  }
+
+  override def next(): ColumnarBatch = {
+    GpuSemaphore.acquireIfNecessary(taskContext)
+    withResource(new NvtxWithMetrics("GpuRange", NvtxColor.DARK_GREEN, opTime)) { _ =>
+      val start = number
+      val remainingRows = (safePartitionEnd - start) / step
+      // Start is inclusive so we need to produce at least one row
+      val rowsExpected = Math.max(1, Math.min(remainingRows, maxRowCountPerBatch))
+      val iter = withRetry(AutoCloseableLong(rowsExpected), splitRowsNumberInHalf) { rows =>
+        withResource(Scalar.fromLong(start)) { startScalar =>
+          withResource(Scalar.fromLong(step)) { stepScalar =>
+            withResource(
+                cudf.ColumnVector.sequence(startScalar, stepScalar, rows.value.toInt)) { vec =>
+              withResource(new Table(vec)) { tab =>
+                GpuColumnVector.from(tab, Array[DataType](LongType))
+              }
+            }
+          }
+        }
+      }
+      assert(iter.hasNext)
+      closeOnExcept(iter.next()) { batch =>
+        // it should has only one batch
+        assert(iter.isEmpty)
+        val endInclusive = start + ((batch.numRows() - 1) * step)
+        number = endInclusive + step
+        if (number < endInclusive ^ step < 0) {
+          // we have Long.MaxValue + Long.MaxValue < Long.MaxValue
+          // and Long.MinValue + Long.MinValue > Long.MinValue, so iff the step causes a
+          // step back, we are pretty sure that we have an overflow.
+          done = true
+        }
+        if (batch.numRows() < rowsExpected) {
+          logDebug(s"Retried with ${batch.numRows()} rows when expected $rowsExpected rows")
+        }
+        batch
+      }
+    }
+  }
+
+  /**
+   * Reduce the input rows number by half, and it returns a Seq with only one element,
+   * that is the half value.
+   */
+  private def splitRowsNumberInHalf: AutoCloseableLong => Seq[AutoCloseableLong] =
+    (rowsNumber) => {
+      withResource(rowsNumber) { _ =>
+        assert(rowsNumber.value > 0)
```

Comment:
Could we change this code a bit? Just so we don't loop forever trying to split something that can never be split.

I normally would pick a number much larger than 10, but to avoid changing the test 10 feels okay.
```
if (rowsNumber.value < 10) {
  throw new SplitAndRetryOOM(s"GPU OutOfMemory: the number of rows generated is too small to be split ${rowsNumber.value}!")
}
Seq(AutoCloseableLong(rowsNumber.value / 2))
```

Reply:
Good suggestion. Updated.